numba/tests/test_struct_ref.py

```diff
@@ -0,0 +1,273 @@
+
+
+@njit
+def ctor_by_intrinsic(vs, ctr):
+    st = my_struct(vs, counter=ctr)
+    st.values += st.values
+    st.counter *= ctr
+    return st
+
+
+@njit
+def ctor_by_class(vs, ctr):
+    return MyStruct(values=vs, counter=ctr)
+
+
+@njit
+def get_values(st):
+    return st.values
+
+
+@njit
+def set_values(st, val):
+    st.values = val
+
+
+@njit
+def get_counter(st):
+    return st.counter
+
+
+@njit
+def compute_fields(st):
+    return st.values + st.counter
+
+
+class TestStructRefBasic(MemoryLeakMixin, TestCase):
+    def test_MySimplerStructType(self):
+        vs = np.arange(10, dtype=np.intp)
+        ctr = 13
+
+        first_expected = vs + vs
+        first_got = ctor_by_intrinsic(vs, ctr)
+        # the returned instance is a structref.StructRefProxy
+        # but not a MyStruct
+        self.assertNotIsInstance(first_got, MyStruct)
+        self.assertPreciseEqual(first_expected, get_values(first_got))
+
+        second_expected = first_expected + (ctr * ctr)
+        second_got = compute_fields(first_got)
+        self.assertPreciseEqual(second_expected, second_got)
+
+    def test_MySimplerStructType_wrapper_has_no_attrs(self):
+        vs = np.arange(10, dtype=np.intp)
+        ctr = 13
+        wrapper = ctor_by_intrinsic(vs, ctr)
+        self.assertIsInstance(wrapper, structref.StructRefProxy)
+        with self.assertRaisesRegex(AttributeError, 'values'):
+            wrapper.values
+        with self.assertRaisesRegex(AttributeError, 'counter'):
+            wrapper.counter
+
+    def test_MyStructType(self):
+        vs = np.arange(10, dtype=np.float64)
+        ctr = 11
+
+        first_expected_arr = vs.copy()
+        first_got = ctor_by_class(vs, ctr)
+        self.assertIsInstance(first_got, MyStruct)
+        self.assertPreciseEqual(first_expected_arr, first_got.values)
+
+        second_expected = first_expected_arr + ctr
+        second_got = compute_fields(first_got)
+        self.assertPreciseEqual(second_expected, second_got)
+        self.assertEqual(first_got.counter, ctr)
+
+    def test_MyStructType_mixed_types(self):
+        # structref constructor is generic
+        @njit
+        def mixed_type(x, y, m, n):
+            return MyStruct(x, y), MyStruct(m, n)
+
+        a, b = mixed_type(1, 2.3, 3.4j, (4,))
+        self.assertEqual(a.values, 1)
+        self.assertEqual(a.counter, 2.3)
+        self.assertEqual(b.values, 3.4j)
+        self.assertEqual(b.counter, (4,))
+
+    def test_MyStructType_in_dict(self):
+        td = typed.Dict()
+        td['a'] = MyStruct(1, 2.3)
+        self.assertEqual(td['a'].values, 1)
+        self.assertEqual(td['a'].counter, 2.3)
+        # overwrite
+        td['a'] = MyStruct(2, 3.3)
+        self.assertEqual(td['a'].values, 2)
+        self.assertEqual(td['a'].counter, 3.3)
+        # mutate
+        td['a'].values += 10
+        self.assertEqual(td['a'].values, 12)    # changed
+        self.assertEqual(td['a'].counter, 3.3)  # unchanged
```

Comment:
Perhaps also check a `td['b'] = MyStruct(4, 5.6)` works, as I think it should? Makes me wonder, should a `MyStruct(int8(4), 5.6)` be castable, I'd guess not at the typing layer?

Reply:
It's not casting the fields at the moment. That would be another potential enhancement